src/prefect/workers/base.py

```diff
@@ -1562,6 +1591,267 @@ async def _mark_flow_run_as_cancelled(
+                exc_info=exc,
+            )
+            self._cancellation_polling_task = asyncio.create_task(
+                critical_service_loop(
+                    workload=self._poll_for_cancelling_flow_runs,
+                    interval=poll_seconds,
+                    jitter_range=0.3,
+                ),
+                name="cancellation-polling",
+            )
+            self._cancellation_polling_task.add_done_callback(
+                self._handle_cancellation_polling_done
+            )
+
+    def _handle_cancellation_polling_done(self, task: asyncio.Task[None]) -> None:
+        """Handle when the cancellation polling task completes or fails."""
+        if task.cancelled():
+            return
+        if exc := task.exception():
+            self._logger.error(
+                "Cancellation polling task failed. Pending flow runs will not be "
+                "cancelled automatically until the worker is restarted.",
+                exc_info=exc,
+            )
+
+    async def _poll_for_cancelling_flow_runs(self) -> None:
+        """Poll for cancelling flow runs in this work pool."""
+        cancelling_flow_runs = await self.client.read_flow_runs(
+            flow_run_filter=FlowRunFilter(
+                state=FlowRunFilterState(
+                    type=FlowRunFilterStateType(any_=[StateType.CANCELLING]),
+                ),
+            ),
+            work_pool_filter=WorkPoolFilter(
+                name=WorkPoolFilterName(any_=[self._work_pool_name])
+            ),
+        )
+
+        for flow_run in cancelling_flow_runs:
+            # Skip if we've already processed this flow run
+            if flow_run.id in self._seen_cancelling_flow_run_ids:
+                continue
+            self._seen_cancelling_flow_run_ids.add(flow_run.id)
+            await self._handle_cancelling_flow_run(flow_run)
+
+    async def _consume_cancellation_events(self) -> None:
+        """Process WebSocket events for real-time cancellation detection."""
+        assert self._cancellation_events_subscriber is not None
+
+        async for event in self._cancellation_events_subscriber:
+            try:
+                flow_run_id = UUID(
+                    event.resource["prefect.resource.id"].replace(
+                        "prefect.flow-run.", ""
+                    )
+                )
+
+                # Skip if we've already seen this flow run
+                if flow_run_id in self._seen_cancelling_flow_run_ids:
+                    continue
+                self._seen_cancelling_flow_run_ids.add(flow_run_id)
+
+                # Fetch the flow run to check state and work pool
+                try:
+                    flow_run = await self.client.read_flow_run(flow_run_id)
+                except ObjectNotFound:
+                    continue
+
+                # Only handle flow runs for this work pool
+                if flow_run.work_pool_name != self._work_pool_name:
+                    continue
```

Comment:
Instead of filtering the events here, what do you think of adding a `related` kwarg to the `EventFilter` provided when the events subscriber is set up? That should allow the worker to only recieve events from flow runs in the current work pool.

Reply:
It took some fiddling, but was able to craft the EventFilter to only get cancelling events for its workpool